File: src/auth/regions.ts
export type RegionValue = "cn" | "i18n" | "us" | "eu" | "codebase";

export interface Region {
  value: RegionValue;
  authUrl: string;
  cookieName: string;
  isCodebase: boolean;
}

const AUTH_URLS: Record<RegionValue, string> = {
  cn: "https://cloud.bytedance.net/auth/api/v1/jwt",
  i18n: "https://cloud-i18n.bytedance.net/auth/api/v1/jwt",
  us: "https://cloud-ttp-us.bytedance.net/auth/api/v1/jwt",
  eu: "https://cloud-i18n.tiktok-eu.org/auth/api/v1/jwt",
  codebase: "https://bits.bytedance.net/api/v1/codebase_token",
};

const ALL_REGIONS: RegionValue[] = ["cn", "i18n", "us", "eu", "codebase"];

export function parseRegion(value: string): Region {
  const v = value.trim().toLowerCase() as RegionValue;
  if (!ALL_REGIONS.includes(v)) {
    throw new Error(
      `未知区域: ${value}，可选: ${ALL_REGIONS.join(", ")}`,
    );
  }
  return {
    value: v,
    authUrl: AUTH_URLS[v],
    cookieName: v === "codebase" ? "CAS_SESSION_API" : "CAS_SESSION",
    isCodebase: v === "codebase",
  };
}

export function allRegions(): Region[] {
  return ALL_REGIONS.map((v) => parseRegion(v));
}

File: internal/auth/regions.go
package auth

import (
	"fmt"
	"strings"
)

type Region struct {
	Value      string
	AuthURL    string
	CookieName string
	IsCodebase bool
}

var authURLs = map[string]string{
	"cn":       "https://cloud.bytedance.net/auth/api/v1/jwt",
	"i18n":     "https://cloud-i18n.bytedance.net/auth/api/v1/jwt",
	"us":       "https://cloud-ttp-us.bytedance.net/auth/api/v1/jwt",
	"eu":       "https://cloud-i18n.tiktok-eu.org/auth/api/v1/jwt",
	"codebase": "https://bits.bytedance.net/api/v1/codebase_token",
}

var allRegionValues = []string{"cn", "i18n", "us", "eu", "codebase"}

func ParseRegion(value string) (Region, error) {
	v := strings.ToLower(strings.TrimSpace(value))
	url, ok := authURLs[v]
	if !ok {
		return Region{}, fmt.Errorf("未知区域: %s，可选: %s", value, strings.Join(allRegionValues, ", "))
	}
	cookieName := "CAS_SESSION"
	if v == "codebase" {
		cookieName = "CAS_SESSION_API"
	}
	return Region{Value: v, AuthURL: url, CookieName: cookieName, IsCodebase: v == "codebase"}, nil
}

func (r Region) LoginURL() string {
	if r.IsCodebase {
		return ""
	}
	return strings.TrimSuffix(r.AuthURL, "/jwt") + "/login"
}

func AllRegions() []Region {
	out := make([]Region, 0, len(allRegionValues))
	for _, value := range allRegionValues {
		region, _ := ParseRegion(value)
		out = append(out, region)
	}
	return out
}
